GDScript function definition and get_node shortcut syntax highlighting

Paulb23 opened the pull request:
This PR starts to add the initial GDScript use cases into the `GDScriptSyntaxHighlighter` separating it from the default.

Firstly function definitions which are mentioned in #17873

Secondly the `get_node` `$` shortcut: 
closes #16284
closes #11554

Paulb23 committed:
GDScript highlighter now remembers previous state

Paulb23 committed:
Added GDScript function definition highlighting

Paulb23 committed:
Added GDScript NodePath highlighting

akien-mga commented on modules/gdscript/gdscript_highlighter.cpp:
```diff
@@ -294,6 +304,12 @@ void GDScriptSyntaxHighlighter::_update_cache() {
 	number_color = text_editor->get_color("number_color");
 	member_color = text_editor->get_color("member_variable_color");
+
+#ifdef TOOLS_ENABLED
```
Do we need to compile any SyntaxHighlighter when tools are disabled?

Paulb23 replied:
I guess not. It's only created when tools are enabled. I'll update with it excluded from scons.

akien-mga commented on editor/editor_settings.cpp:
```diff
@@ -583,6 +583,9 @@ void EditorSettings::_load_default_text_editor_theme() {
 	_initial_set("text_editor/highlighting/search_result_border_color", Color(0.1, 0.45, 0.1, 1));
+
+	// GDScript highlighter
+	_initial_set("text_editor/highlighting/GDScript/function_definition_color", Color::html("#01e1ff"));
```
I'm not sure about using uppercase letters in parameter names, we haven't done it so far so it's maybe better to stick to snake_case.

akien-mga commented on editor/editor_settings.cpp:
```diff
@@ -619,6 +622,9 @@ bool EditorSettings::_save_text_editor_theme(String p_file) {
 	cf->set_value(theme_section, "search_result_border_color", ((Color)get("text_editor/highlighting/search_result_border_color")).to_html());
 
+	//GDScript highlighter
+	cf->set_value(theme_section, "GDScript/function_definition_color", ((Color)get("text_editor/highlighting/GDScript/function_definition_color")).to_html());
```
Maybe it should be made possible for SyntaxHighlighters to register such options themselves, so that it can be done also for other languages which are added as GDNative modules.

Paulb23 replied:
Yeah I've been thinking about this but can't arrive at a decent solution. It would need both adding here and under `editor_themes.cpp` for light / dark theme switching.

vnen replied:
While this would be a nice addition, I think it's out of scope of this PR. I believe it'll need a bit of a refactor for EditorSettings to allow modules registering settings in a timely manner.

Paulb23 committed:
Exclude GDScriptSyntaxHighlighter from non-tool builds

Paulb23 commented:
Moved `gdscript_highlighter` into a `editor` subdirectory and updated the `SCsub` to only include this folder on tool builds. 

Lower cased the settings path.

reduz commented:
What's the status on this?

Paulb23 commented:
It's complete.

akien-mga commented:
Thanks!

MarianoGnu commented:
do you mind an screenshot?
